src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -1362,6 +1362,114 @@ static LogicalResult getVariadicSimdCodeFullyFlattened(
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// Utilities for Op fusion at lowering
+//===----------------------------------------------------------------------===//
+
+// Function pointer type for emitScalarOpFor<T>
+typedef Value (*EmitScalarFunc)(ConversionPatternRewriter &rewriter,
+    Location loc, Operation *op, Type elementType,
+    ArrayRef<Value> scalarOperands);
+
+// Variadic template to iterate all the fusible Ops
+template <typename T>
+bool enqueueFusedOpImpl(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions) {
+  if (isa<T>(op)) {
+    fusibleOps.emplace_back(op);
+    fuseEmitFunctions.emplace_back(emitScalarOpFor<T>);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+template <typename T = void, class... Ts>
+bool enqueueFusedOp(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions);
+
+template <typename T, class... Ts>
+bool enqueueFusedOp(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions) {
+  if (enqueueFusedOpImpl<T>(op, fusibleOps, fuseEmitFunctions)) {
+    return true;
+  } else {
+    return enqueueFusedOp<Ts...>(op, fusibleOps, fuseEmitFunctions);
+  }
+}
+
+template <>
+bool enqueueFusedOp(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
+    SmallVector<EmitScalarFunc, 2> &fuseEmitFunctions) {
+  return false;
+}
+
+bool enqueueFusedOp(Operation *op, SmallVector<Operation *, 2> &fusibleOps,
```

Comment:
While this recursive template is very crafty, I am not very fond of this approach because if forces us to look into a list that is separate of the function definition.

I would prefer something that uses the declarative type that we already use. Here is what I used to do for SIMD before I changed it to something else.

```
using SimdScalarOp = std::true_type;    // scalar op can be simdized.
using NoSimdScalarOp = std::false_type; // scalar op cannot be simdized.

template <typename Op>
struct ScalarOp {
  using FOp = NotSuportedScalarOp;
  using IOp = NotSuportedScalarOp;
  using SimdEnabled = NoSimdScalarOp;
};
```

So you would use that pattern for `FusionEnabled` with `FusibleOp` and `NonFusibleOp`.

Then you can use a simple if..

```C++
    if constexpr (SimdizableOp<ElementwiseUnaryOp>::value) {
      // SIMD is enabled for this operation, test if desired and feasible

```

This code let you have an "if" that will statically be enabled when the op can be fused.

I prefer this approach because when we look at the function, we can clearly see declaratively if it can be fused. If we add a new operation, we will be forced to define it one way or another, so it is also better for maintainability.

Reply:
I feel the difference is whether to instantiate/specialize the template manually or not. The SIMD/scalar code emission needs to define the llvm op for each ONNX Op, which has to be done manually. Here, it's just true or false based on the list of ONNX Op types specified. 
The best way, to me, is to use traits/interface for dialect.

Reply:
SIMD flag was only defined for the broadcast ops.

Reply:
Anyway, can we have slight different way of programming?

Reply:
simd or all the element-wise emit function is called from the Op with its ONNX Op type known.   The extra work for fusion detection is to check the type of a generic op, which is the user of the current Op.

Reply:
It looks using traits/interface is better especially when we extend the fusion for other ops like Conv rather than element-wise ops. Currently `struct ScalarOp` is internal to element-wise ops only.
Perhaps, we can take the same approach as `shapeHelper`. Then, check if an op has a trait, like `UnaryOp` then get its `emitScalarFunc` to fuse.

Reply:
Yes. The final solution is to use interface. We can define an interface for krnl lowering.
You are right on the struct ScalarOp is to element-wise ops only (with op type known) and we need the interface to check whether an ONNX Op has the interface or not.